internal/testing/exchange/exchange.go

```diff
@@ -0,0 +1,159 @@
+package exchange
+
+import (
+	"errors"
+	"fmt"
+	"log"
+	"net/http"
+	"net/http/httptest"
+	"strings"
+	"sync"
+	"testing"
+
+	"github.com/gorilla/websocket"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"github.com/thrasher-corp/gocryptotrader/config"
+	exchange "github.com/thrasher-corp/gocryptotrader/exchanges"
+	"github.com/thrasher-corp/gocryptotrader/exchanges/mock"
+	"github.com/thrasher-corp/gocryptotrader/exchanges/sharedtestvalues"
+	"github.com/thrasher-corp/gocryptotrader/exchanges/subscription"
+)
+
+// TestInstance takes an empty exchange instance and loads config for it from testdata/configtest and connects a NewTestWebsocket
+func TestInstance(e exchange.IBotExchange) error {
+	cfg := config.GetConfig()
+	err := cfg.LoadConfig("../../testdata/configtest.json", true)
+	if err != nil {
+		return fmt.Errorf("LoadConfig() error: %w", err)
+	}
+	parts := strings.Split(fmt.Sprintf("%T", e), ".")
+	if len(parts) != 2 {
+		return errors.New("unexpected parts splitting exchange type name")
+	}
+	eName := parts[1]
+	exchConf, err := cfg.GetExchangeConfig(eName)
+	if err != nil {
+		return fmt.Errorf("GetExchangeConfig(`%s`) error: %w", eName, err)
+	}
+	e.SetDefaults()
+	b := e.GetBase()
+	b.Websocket = sharedtestvalues.NewTestWebsocket()
+	err = e.Setup(exchConf)
+	if err != nil {
+		return fmt.Errorf("Setup() error: %w", err)
+	}
+	return nil
+}
+
+// httpMockFile is a consistent path under each exchange to find the mock server definitions
+const httpMockFile = "testdata/http.json"
+
+// MockHTTPInstance takes an existing Exchange instance and attaches it to a new http server
+// It is expected to be run once,  since http requests do not often tangle with each other
+func MockHTTPInstance(e exchange.IBotExchange) error {
+	serverDetails, newClient, err := mock.NewVCRServer(httpMockFile)
+	if err != nil {
+		return fmt.Errorf("mock server error %s", err)
+	}
+	b := e.GetBase()
+	b.SkipAuthCheck = true
+	err = b.SetHTTPClient(newClient)
+	if err != nil {
+		return fmt.Errorf("mock server error %s", err)
+	}
+	endpointMap := b.API.Endpoints.GetURLMap()
+	for k := range endpointMap {
+		err = b.API.Endpoints.SetRunning(k, serverDetails)
+		if err != nil {
+			return fmt.Errorf("mock server error %s", err)
+		}
+	}
+	log.Printf(sharedtestvalues.MockTesting, e.GetName())
+
+	return nil
+}
+
+var upgrader = websocket.Upgrader{}
+
+type wsMockFunc func(msg []byte, w *websocket.Conn) error
+
+// MockWSInstance creates a new Exchange instance with a mock WS instance and HTTP server
+// It accepts an exchange package type argument and a mock WS function
+// It is expected to be run from any WS tests which need a specific response function
+func MockWSInstance[T any, PT interface {
```

Comment:
Same as above. This is because the branch for mock testing was brought forwards from under the Kraken PR, and it includes several shared testing features.